### ISSUE TITLE
ARM NEON intrinsics are unsound due to subnormal flushing

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#129880]
This is the ARM NEON version of https://github.com/rust-lang/rust/issues/114479. Example by @beetrees, to be compiled with `--target armv7-unknown-linux-gnueabihf -O -Ctarget-feature=+neon`:
```rust
#![feature(stdarch_arm_neon_intrinsics)]

use std::arch::arm::{float32x2_t, vadd_f32};
use std::mem::transmute;

#[inline(never)]
fn print_vals(x: float32x2_t, i: usize, vals_i: u32) {
    println!("x={x:?} i={i} vals[i]={vals_i}");
}

const ZERO: float32x2_t = unsafe { transmute([0, 0]) };
const INC: float32x2_t = unsafe { transmute([f32::MIN_POSITIVE / 128.0, f32::MIN_POSITIVE / 128.0]) };
const TARGET: [u32; 2] = unsafe { transmute([f32::MIN_POSITIVE, f32::MIN_POSITIVE]) };

#[inline(never)]
pub fn evil(vals: &[u32; 300]) {
    let mut x: float32x2_t = ZERO;
    let mut i: usize = 0;
    while unsafe { transmute::<float32x2_t, [u32; 2]>(x) } != TARGET {
        print_vals(x, i, vals[i]);
        x = unsafe { vadd_f32(x, INC) };
        x = unsafe { vadd_f32(x, INC) };
        i += 2;
    }
}

pub fn main() {
    let mut vals: [u32; 300] = [0; 300];
    for i in 0..300 { vals[i as usize] = i; }
    evil(&vals);
}

#[cfg(not(target_feature = "neon"))]
compile_error!("-Ctarget-feature=+neon required");
```

LLVM's optimizations assume they can calculate what that loop does, and that it follows IEEE semantics. But LLVM's codegen produces code that does not have IEEE semantics, and instead flushes subnormals to zero. :boom: 

This almost surely also affects the unstable `std::simd` on ARM.

**Comment by apiraino** [comment]
WG-prioritization assigning priority ([Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.23129880.20ARM.20NEON.20intrinsics.20are.20unsound.20due.20to.20subnormal.20fl.E2.80.A6)).

@rustbot label -I-prioritize +P-medium

**Comment by RalfJung** [comment]
@nikic IIRC you said that scalar evolution stopped evaluating float-returning operations because of determinism issues... shouldn't that also avoid this problem?

**Comment by nikic** [comment]
@RalfJung No, constant folding distinguishes whether the operation is non-deterministic or not, and the primitive ones (without NaNs and FMF flags) are deterministic.

It can be disabled by specifying `"denormal-fp-math"="dynamic,dynamic"`, but that means disabling *all* constant folding on denormals.

**Comment by RalfJung** [comment]
Constant folding is fine even for non-deterministic operations, that's just a refinement; the issue is assuming it gives a particular result without actually enforcing that this is the choice that will be made later -- e.g., scalar evolution.

But I think I then just misunderstood what you said [here](https://github.com/llvm/llvm-project/pull/172012#issuecomment-3648125264). I now realize you said "all FP libcalls/intrinsics". I didn't realize that there is some sense in which fadd is "more primitive" than minnum.
